### PR TITLE
addindex/disttask: adjust add index task concurrency & add check when submit task

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #49008

Problem Summary:

### What changed and how does it work?
- adjust add index task concurrency to make sure it can be scheduled
- add sanity check that task.Concurrency should < cpu-cpunt when submit task

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/backfilling_operators.go [hunk 1]
```diff
@@ -131,17 +130,19 @@ func getWriterMemSize(idxNum int) (uint64, error) {
...
-func getMergeSortPartSize(concurrency int, idxNum int) (uint64, error) {
-	writerMemSize, err := getWriterMemSize(idxNum)
+func getMergeSortPartSize(workerCntLimit, idxNum int) (uint64, error) {
+	_, writerCnt := calculateReadWriteWorkerCnt(workerCntLimit)
+	writerMemSize, err := getWriterMemSize(writerCnt, idxNum)
 	if err != nil {
 		return 0, nil
 	}
-	return writerMemSize / uint64(concurrency) / 10, nil
+	return writerMemSize / uint64(workerCntLimit) / 10, nil
```
Comment (D3Hunter) [inline pkg/ddl/backfilling_operators.go:139]: @ywqzzy should use taskConcurrency? as this worker-cnt might > cpu-count

Comment (ywqzzy) [reply pkg/ddl/backfilling_operators.go:139]: Yes

File: tests/realtikvtest/addindextest1/disttask_test.go [hunk 2]
```diff
@@ -61,6 +77,18 @@ func TestAddIndexDistBasic(t *testing.T) {
...
 	tk.MustExec("admin check index t idx;")
+	taskMgr, err := storage.GetTaskManager()
+	require.NoError(t, err)
+	ctx := util.WithInternalSourceType(context.Background(), "dispatcher")
+	task, err := taskMgr.GetTaskByIDWithHistory(ctx, storage.TestLastTaskID.Load())
+	require.NoError(t, err)
+	require.Equal(t, 1, task.Concurrency)
+	taskMeta := ddl.BackfillTaskMeta{}
+	require.NoError(t, json.Unmarshal(task.Meta, &taskMeta))
+	require.Equal(t, 111, taskMeta.WorkerCntLimit)
```
Comment (ywqzzy) [inline tests/realtikvtest/addindextest1/disttask_test.go:88]: Why setting workerCntLimit > cpu count? If task concurrency != workerCnt, the reserved resource is less than real consuming resource.

File: pkg/disttask/framework/proto/task.go [hunk 3]
```diff
@@ -24,23 +24,19 @@ import (
...
-//		┌──────┐          ├────────►│cancelling│
-//		│failed│          │         └────┬─────┘
-//		└──────┘          │              ▼
-//		                  │         ┌─────────┐     ┌────────┐
-//		                  └────────►│reverting├────►│reverted│
-//		                            └────┬────┘     └────────┘
-//		                                 │          ┌─────────────┐
-//		                                 └─────────►│revert_failed│
-//		                                            └─────────────┘
+//		└──┬────┘     └──┬┬───┘     └────────┘
```
Comment (ywqzzy) [inline pkg/disttask/framework/proto/task.go:33]: when will the task state transform from pending to failed?

File: pkg/ddl/backfilling_merge_sort.go [hunk 4]
```diff
@@ -41,21 +40,23 @@ type mergeSortExecutor struct {
 	cloudStoreURI       string
 	mu                  sync.Mutex
 	subtaskSortedKVMeta *external.SortedKVMeta
+	workerCntLimit      int
```
Comment (ywqzzy) [inline pkg/ddl/backfilling_merge_sort.go:43]: In local add index, the workerCnt can be adjusted dynamiclly.
Need to discuss whether we will support the feature.

Comment (ywqzzy) [reply pkg/ddl/backfilling_merge_sort.go:43]: > In local add index, the workerCnt can be adjusted dynamically. Need to discuss whether we will support the feature.

ok for now

**Review by tangenta — APPROVED** [review]

File: pkg/disttask/framework/proto/task.go (hunk 3 above)
Comment (D3Hunter) [reply pkg/disttask/framework/proto/task.go:33]: invalid task type or dispatcher init failed

File: pkg/ddl/backfilling_merge_sort.go (hunk 4 above)
Comment (D3Hunter) [reply pkg/ddl/backfilling_merge_sort.go:43]: > In local add index, the workerCnt can be adjusted dynamically.

is this an exposed feature or just previous code can support use newest workerCnt for later subtask?

Comment (ywqzzy) [reply pkg/ddl/backfilling_merge_sort.go:43]: @tangenta cc

File: tests/realtikvtest/addindextest1/disttask_test.go (hunk 2 above)
Comment (D3Hunter) [reply tests/realtikvtest/addindextest1/disttask_test.go:88]: 1 add-index worker might not consume 1 core, so `GetDDLReorgWorkerCounter` can be larger than core-count, see https://docs.pingcap.com/tidb/stable/system-variables#tidb_ddl_reorg_worker_cnt.

if user do set tidb_ddl_reorg_worker_cnt > core-count, we reserve resource using core-count, and that's what we can reserve the most.

File: pkg/ddl/backfilling_merge_sort.go (hunk 4 above)
Comment (D3Hunter) [reply pkg/ddl/backfilling_merge_sort.go:43]: as it's an exposed behavior, i will revert the part that use saved worker-count
https://docs.pingcap.com/tidb/stable/ddl-introduction#balance-the-physical-ddl-execution-speed-and-the-impact-on-application-load-through-system-variables

**Review by ywqzzy — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Pull request merged**